alacritty_terminal/src/grid/mod.rs

```diff
@@ -665,6 +665,33 @@ impl<T> Grid<T> {
     }
 }
 
+impl<T: GridCell + ToString> Grid<T> {
+    fn to_string_between(&self, y0: Line, y1: Line) -> String {
+        let mut s = String::new();
+        for y in (*y0..*y1).rev() {
+            let row = &self[y];
+            for x in 0..*self.num_cols() {
+                let cell = &row[Column(x)];
+                s.push_str(&cell.to_string()); // TODO: this is not cheap, support for Cell only?
```

Comment:
Passing a `&mut String` like `s` in this case and then writing to it is **a lot** cheaper than calling `to_string()`.

Since both the primary cell content and the zerowidth chars are accessible, you can just do all the conversion out here rather than inside the cell's methods.

You can see how I've done that here:
https://github.com/chrisduerr/alacritty/blob/pager/alacritty_terminal/src/term/mod.rs#L2152-L2178

Though your implementation is much, much more trivial, since you're not trying to handle escapes.

Reply:
In this case the cell content is actually not accessible because I'm only assuming `GridCell`. Hence the need for `ToString`, and implementing it for `Cell`, to be able to do this at all here. Can/should I extend `GridCell`, or instead impl only for `Grid<Cell>` here as the comment suggests?

Reply:
Could be implemented on top of term, where you know the type.

There are a couple of options available, just a matter of comparing to see which is the cleanest one.

Reply:
Sure. I pushed a commit that moves it on top of `Term` since that seemed the most straightforward — I can squash it later, figured it's better to have the options open for now. I'll leave it up to you to determine what's cleanest as I'm not really familiar with the codebase and what sort of abstraction boundaries you prefer to keep.